render/pixman/renderer.c

```diff
@@ -0,0 +1,365 @@
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+	struct wlr_pixman_buffer *buffer = renderer->current_buffer;
+
+	struct pixman_color mask_colour = {0};
+	mask_colour.alpha = 0xFFFF * alpha;
+	pixman_image_t *mask = pixman_image_create_solid_fill(&mask_colour);
+
+	struct pixman_transform transform = {0};
+	wlr_matrix_scale((float*)matrix, 1.0 / fbox->width, 1.0 / fbox->height);
+	matrix_to_pixman_transform(&transform, matrix);
+	pixman_transform_invert(&transform, &transform);
+
+	pixman_image_set_transform(texture->image, &transform);
+
+	pixman_image_composite32(PIXMAN_OP_OVER, texture->image, mask,
+			buffer->image, 0, 0, 0, 0, 0, 0, renderer->width,
+			renderer->height);
+
+	pixman_image_unref(mask);
+
+	return true;
+}
+
+static void pixman_render_quad_with_matrix(struct wlr_renderer *wlr_renderer,
+		const float color[static 4], const float matrix[static 9]) {
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+	struct wlr_pixman_buffer *buffer = renderer->current_buffer;
+
+	struct pixman_color colour = {
+		.red = color[0] * 0xFFFF,
+		.green = color[1] * 0xFFFF,
+		.blue = color[2] * 0xFFFF,
+		.alpha = color[3] * 0xFFFF,
+	};
+
+	pixman_image_t *fill = pixman_image_create_solid_fill(&colour);
+	pixman_image_t *image = pixman_image_create_bits (PIXMAN_a8r8g8b8,
+						matrix[0], matrix[4], NULL, 0);
+	pixman_image_composite32 (PIXMAN_OP_SRC, fill, NULL, image,
+				  0, 0, 0, 0, 0, 0, matrix[0], matrix[4]);
+	pixman_image_unref(fill);
+
+	struct pixman_transform transform = {0};
+	wlr_matrix_scale((float*)matrix, 1.0 / matrix[0], 1.0 / matrix[4]);
+	matrix_to_pixman_transform(&transform, matrix);
+	pixman_transform_invert(&transform, &transform);
+
+	pixman_image_set_transform(image, &transform);
+
+	pixman_image_composite32(PIXMAN_OP_OVER, image, NULL,
+			buffer->image, 0, 0, 0, 0, 0, 0, renderer->width,
+			renderer->height);
+
+	pixman_image_unref(image);
+}
+
+static const uint32_t *pixman_get_shm_texture_formats(
+		struct wlr_renderer *wlr_renderer, size_t *len) {
+	return get_pixman_drm_formats(len);
+}
+
+static const struct wlr_drm_format_set *pixman_get_render_formats(
+		struct wlr_renderer *wlr_renderer) {
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+	return &renderer->drm_formats;
+}
+
+static struct wlr_texture *pixman_texture_from_pixels(
+		struct wlr_renderer *wlr_renderer, uint32_t drm_format,
+		uint32_t stride, uint32_t width, uint32_t height, const void *data) {
+	struct wlr_pixman_renderer *renderer = get_renderer(wlr_renderer);
+	struct wlr_pixman_texture *texture =
+		calloc(1, sizeof(struct wlr_pixman_texture));
+	if (texture == NULL) {
+		wlr_log(WLR_ERROR, "Failed to allocate pixman texture");
+		return NULL;
+	}
+
+	wlr_texture_init(&texture->wlr_texture, &texture_impl, width, height);
+	texture->renderer = renderer;
+
+	texture->format = drm_get_pixel_format_info(drm_format);
+	assert(texture->format);
```

Comment:
Can we turn this assert into a graceful failure? I'd prefer not to die when a format which isn't in our DRM pixel table is passed.